# Tâche 3 - Corriger des proverbes avec des *transformers*

Tout comme pour le premier travail pratique, l'objectif de cette tâche est de corriger des proverbes. Cependant vous devrez choisir et utiliser des modèles transformers (2) pour identifier le mot à remplacer et choisir le meilleur mot à insérer en fonction du contexte du proverbe.


Consignes :
- Corriger un proverbe consiste à remplacer un verbe par l’un des mots proposés dans la liste.
- Choix des modèles transformer: Vous devrez choisir 2 modèles, un pour faire l’analyse grammaticale (*POS tagging*) et l'autre pour choisir le mot à insérer dans le proverbe. Deux contraintes : a) pas de LLMs, et b) utilisez des modèles de HuggingFace qui peuvent traiter le français (ou qui sont multilingues).
- Entraînement des modèles : Vous pouvez utiliser les modèles préentraînés sans modification. Me consulter en cas de doute.  
- Segmentation : Ne pas segmenter un proverbe en sous-phrases.
- Tokenisation et plongements de mots : Ceux des modèles utilisés.
- Prétraitement et normalisation : Rien à faire.
- Code : Utilisez le fichier ***t3_corriger_proverbes.ipynb*** (ce fichier) pour mener vos expérimentations.
- Données : Évaluez la performance à l’aide du fichier ***data/t3_test_proverbes.json*** et faire l'analyse de vos résultats.
- Question : Comment se compare l’approche retenue pour cette tâche par rapport à celle du Travail pratique #1?

Vous pouvez ajouter au *notebook* toutes les cellules dont vous avez besoin pour votre code, vos explications ou la présentation de vos résultats. Vous pouvez également ajouter des sous-sections (par ex. des sous-sections 1.1, 1.2 etc.) si cela améliore la lisibilité.

Notes :
- Évitez les bouts de code trop longs ou trop complexes. Par exemple, il est difficile de comprendre 4-5 boucles ou conditions imbriquées. Si c'est le cas, définissez des sous-fonctions pour refactoriser et simplifier votre code.
- Expliquez sommairement votre démarche.
- Expliquez les choix que vous faites au niveau de la programmation et des modèles (si non trivial).
- Analyser vos résultats. Indiquez ce que vous observez, si c'est bon ou non, si c'est surprenant, etc.
- Une analyse quantitative et qualitative d'erreurs est intéressante et permet de mieux comprendre le comportement d'un modèle.

## Section 1 - Lecture du fichier de test

Le fichier de test ***./data/t3_test_proverbes.json*** contient les proverbes modifiés, la liste de mots candidats et la bonne version du proverbe.

In [1]:
import json

def load_tests(filename):
    with open(filename, 'r', encoding='utf-8') as fp:
        test_data = json.load(fp)
    return test_data

In [3]:
#test_fn = './data/t3_test_proverbes.json'  # Le fichier de test = À modifier selon votre configuration

test_fn = 'data/t3_test_proverbes.json'

In [4]:
tests = load_tests(test_fn)

In [5]:
import pandas as pd

def get_dataframe(test_proverbs):
    return pd.DataFrame.from_dict(test_proverbs, orient='columns', dtype=None, columns=None)

df = get_dataframe(tests)
df

Masked               Word_list  \
0                      a beau mentir qui part de loin        [vient, revient]   
1                     a beau dormir qui vient de loin        [partir, mentir]   
2                          l’occasion forge le larron        [fait, occasion]   
3                        endors-toi, le ciel t’aidera           [bouge, aide]   
4                            aide-toi, le ciel t’aura          [aidera, aide]   
5                      ce que femme dit, dieu le veut             [dit, veut]   
6                 ce que femme veut, dieu le souhaite             [dit, veut]   
7                      bien mal acquis ne sait jamais         [profite, fait]   
8               bon ouvrier ne déplace pas ses outils        [fait, querelle]   
9            pour le fou, c’était tous les jours fête               [est, es]   
10                          dire et plaire, sont deux           [dire, faire]   
11                         manger et faire, sont deux           [dire, faire]   
12                     mieux vaut prévenir que courir      [prévenir, guérir]   
13                       mieux vaut dormir que guérir      [prévenir, guérir]   
14                  à qui dieu aide, nul ne peut être           [aide, nuire]   
15                 à qui dieu veut, nul ne peut nuire           [aide, nuire]   
16                    il faut le faire pour le croire           [voir, boire]   
17                     il faut le voir pour le saisir         [boire, croire]   
18  on ne mord pas le poisson qui est encore dans ...             [vend, dit]   
19  on ne vend pas le poisson qui navigue encore d...             [est, nage]   
20                       le poisson mange par la tête      [pourrit, respire]   
21  repose-toi plutôt sans souper, que de te lever...          [lève, couche]   
22  couche-toi plutôt sans souper, que de te retro...         [mettre, lever]   
23  les montagnes ne se déplacent point, mais les ...   [rencontrent, paient]   
24  les montagnes ne se rencontrent point, mais le...  [rencontrent, battent]   
25           étudier peu, chasse beaucoup de maladies         [manger, parle]   

                                              Proverb  
0                     a beau mentir qui vient de loin  
1                     a beau mentir qui vient de loin  
2                           l’occasion fait le larron  
3                          aide-toi, le ciel t’aidera  
4                          aide-toi, le ciel t’aidera  
5                     ce que femme veut, dieu le veut  
6                     ce que femme veut, dieu le veut  
7                   bien mal acquis ne profite jamais  
8              bon ouvrier ne querelle pas ses outils  
9              pour le fou, c’est tous les jours fête  
10                           dire et faire, sont deux  
11                           dire et faire, sont deux  
12                     mieux vaut prévenir que guérir  
13                     mieux vaut prévenir que guérir  
14                 à qui dieu aide, nul ne peut nuire  
15                 à qui dieu aide, nul ne peut nuire  
16                     il faut le voir pour le croire  
17                     il faut le voir pour le croire  
18  on ne vend pas le poisson qui est encore dans ...  
19  on ne vend pas le poisson qui est encore dans ...  
20                     le poisson pourrit par la tête  
21  couche-toi plutôt sans souper, que de te lever...  
22  couche-toi plutôt sans souper, que de te lever...  
23  les montagnes ne se rencontrent point, mais le...  
24  les montagnes ne se rencontrent point, mais le...  
25            manger peu, chasse beaucoup de maladies

## Section 2 - Code pour repérer les mots qui pourraient être remplacés dans un proverbe modifié

Expliquez ici comment vous procédez pour identifier les mots d'un proverbe qui pourraient faire l'objet d'une substitution.  



In [6]:
#utile pour s'assurer que ce token est un verbe et non pas reparti sur plusieurs tokens pour verifier la nouvelle condition de 2 verbes (non reparti sur plusieurs tokens) et succeffifs
def verifier_verbe(verbe):
    # Liste des terminaisons valides
    terminaisons = ["er", "ir", "oir", "re"]
    # Vérifie si le verbe se termine par l'une des terminaisons
    return any(verbe.endswith(terminaison) for terminaison in terminaisons)
print(verifier_verbe("peut"))

# Exemple d'utilisation
verbes = ["manger", "finir", "voir", "prendre", "courir","vaut","peut"]
resultats = {verbe: verifier_verbe(verbe) for verbe in verbes}

# Affiche les résultats
for verbe, est_valide in resultats.items():
    print(f"Le verbe '{verbe}' a une terminaison valide: {est_valide}")

False
Le verbe 'manger' a une terminaison valide: True
Le verbe 'finir' a une terminaison valide: True
Le verbe 'voir' a une terminaison valide: True
Le verbe 'prendre' a une terminaison valide: True
Le verbe 'courir' a une terminaison valide: True
Le verbe 'vaut' a une terminaison valide: False
Le verbe 'peut' a une terminaison valide: False


In [7]:
#derniere version corrigee
def process_verbs_and_other_tokens(tokens):
    """
    Traite une liste de tokens, en fusionnant ou séparant les verbes selon des règles précises.

    - Concatène certains tokens pour former des verbes (par exemple, "en dor s" -> "endors").
    - Sépare les verbes consécutifs comme "peut" et "faire" si le deuxième verbe satisfait `verifier_verbe`.

    Args:
        tokens (list of tuple): Liste de (token, pos_tag) tuples.

    Returns:
        list of tuple: Liste des tokens traités.
    """
    result = []
    verb_sequence = []  # Pour collecter les morceaux d'un seul verbe complexe

    for i, (token, pos_tag) in enumerate(tokens):
        if pos_tag == "VERB":
            if verb_sequence:
                # Vérifie si le token courant doit démarrer une nouvelle séquence
                if verifier_verbe(token):
                    # Ajoute la séquence précédente comme un verbe complet
                    result.append(("".join(verb_sequence), "VERB"))
                    verb_sequence = []
                    result.append((token, "VERB"))  # Nouveau verbe validé
                else:
                    # Continue la concaténation pour former un seul verbe complexe
                    verb_sequence.append(token)
            else:
                # Démarre une nouvelle séquence pour un verbe potentiel
                verb_sequence.append(token)
        else:
            # Gère les cas où un non-verbe suit une séquence
            if verb_sequence:
                result.append(("".join(verb_sequence), "VERB"))
                verb_sequence = []
            result.append((token, pos_tag))

    # Ajoute les verbes restants
    if verb_sequence:
        result.append(("".join(verb_sequence), "VERB"))

    return result


# Exemple d'utilisation
tokens_1 = [
    ("en", "VERB"), ("dor", "VERB"), ("s", "VERB"),
    ("-", "VERB"), ("toi", "PRON"), ("le", "DET"),
    ("peut", "VERB"), ("faire", "VERB"),
    ("ciel", "NC"), ("t", "CLO"), ("aide", "VERB"), ("ra", "VERB")
]

output_1 = process_verbs_and_other_tokens(tokens_1)
print("Résultat 1:", output_1)


Résultat 1: [('endors-', 'VERB'), ('toi', 'PRON'), ('le', 'DET'), ('peut', 'VERB'), ('faire', 'VERB'), ('ciel', 'NC'), ('t', 'CLO'), ('aidera', 'VERB')]


In [8]:
#the latest code found with more postags for more  details  like -t is recognized i hope it could help in the endors-toi et ..

from transformers import CamembertTokenizer, CamembertForTokenClassification, TokenClassificationPipeline

tokenizer = CamembertTokenizer.from_pretrained('qanastek/pos-french-camembert')
model = CamembertForTokenClassification.from_pretrained('qanastek/pos-french-camembert')
pos = TokenClassificationPipeline(model=model, tokenizer=tokenizer)

def make_prediction(sentence):
    labels = [l['entity'] for l in pos(sentence)]
    return list(zip(sentence.split(" "), labels))

res = make_prediction("George Washington est allé à Washington")

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/455 [00:00<?, ?B/s]

sentencepiece.bpe.model:   0%|          | 0.00/811k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/299 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/3.20k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/440M [00:00<?, ?B/s]

In [12]:
#J'ai rendu la fonction précédente comme fonction: pour l'utiliser comme résultat du 1er modèle
def result_model1(df):
  df1 = df.copy()
  df1['Postags'] = [[] for _ in range(len(df))]
  for i in range(len(df)):
    tokens_after_posttraitement = []
    text = df['Masked'][i]
    result = pos(text)

    print("Example of proverb{}\n".format(i))

    # Display results with better formatting
    for entity in result:
        print(f"Token: {entity['word']}, Tag: {entity['entity']}, Score: {entity['score']:.6f}")
        tokens_after_posttraitement.append((entity['word'], entity['entity']))
    print("\n")
    df1['Postags'][i] = tokens_after_posttraitement
    print(process_verbs_and_other_tokens(tokens_after_posttraitement))
    print("The form for the verb in the wordlist:\n")
    for j in range(2):
      verb = df['Word_list'][i][j]
      result_verb = pos(verb)
      print(result_verb)
      print("\n")

  return df1

In [13]:
df1=result_model1(df)

Example of proverb0

Token: ▁a, Tag: VERB, Score: 0.991083
Token: ▁beau, Tag: ADJMS, Score: 0.998540
Token: ▁mentir, Tag: VERB, Score: 0.999797
Token: ▁qui, Tag: PREL, Score: 0.999337
Token: ▁part, Tag: VERB, Score: 0.999828
Token: ▁de, Tag: PREP, Score: 0.999889
Token: ▁loin, Tag: ADV, Score: 0.999752


[('▁a', 'VERB'), ('▁beau', 'ADJMS'), ('▁mentir', 'VERB'), ('▁qui', 'PREL'), ('▁part', 'VERB'), ('▁de', 'PREP'), ('▁loin', 'ADV')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.9975605, 'index': 1, 'word': '▁vient', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.998572, 'index': 1, 'word': '▁revient', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb1

Token: ▁a, Tag: VERB, Score: 0.998252
Token: ▁beau, Tag: ADJMS, Score: 0.998601
Token: ▁dormir, Tag: VERB, Score: 0.999767
Token: ▁qui, Tag: PREL, Score: 0.999450
Token: ▁vient, Tag: VERB, Score: 0.999811
Token: ▁de, Tag: PREP, Score: 0.999890
Token: ▁loin, Tag: ADV, Score: 0.999738


[('▁a', 'VERB'), ('▁beau', 'ADJMS'), ('▁dormir', 'VERB'), ('▁qui', 'PREL'), ('▁vient', 'VERB'), ('▁de', 'PREP'), ('▁loin', 'ADV')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.99976856, 'index': 1, 'word': '▁partir', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.9990521, 'index': 1, 'word': '▁mentir', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb2

Token: ▁l, Tag: DET, Score: 0.999735
Token: ’, Tag: ADV, Score: 0.670996
Token: occasion, Tag: NFS, Score: 0.999490
Token: ▁forge, Tag: VERB, Score: 0.999789
Token: ▁le, Tag: DETMS, Score: 0.999836
Token: ▁la, Tag: NMS, Score: 0.999415
Token: r, Tag: NMS, Score: 0.999519
Token: ron, Tag: NMS, Score: 0.998672


[('▁l', 'DET'), ('’', 'ADV'), ('occasion', 'NFS'), ('▁forge', 'VERB'), ('▁le', 'DETMS'), ('▁la', 'NMS'), ('r', 'NMS'), ('ron', 'NMS')]
The form for the verb in the wordlist:

[{'entity': 'NMS', 'score': 0.9157817, 'index': 1, 'word': '▁fait', 'start': None, 'end': None}]


[{'entity': 'NFS', 'score': 0.99847966, 'index': 1, 'word': '▁occasion', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb3

Token: ▁en, Tag: VERB, Score: 0.999724
Token: dor, Tag: VERB, Score: 0.999818
Token: s, Tag: VERB, Score: 0.999823
Token: -, Tag: VERB, Score: 0.999813
Token: toi, Tag: PPOBJMS, Score: 0.997650
Token: ,, Tag: PUNCT, Score: 0.999907
Token: ▁le, Tag: DETMS, Score: 0.999869
Token: ▁ciel, Tag: NMS, Score: 0.999762
Token: ▁t, Tag: PREFS, Score: 0.990742
Token: ’, Tag: ADV, Score: 0.154792
Token: aider, Tag: VERB, Score: 0.999794
Token: a, Tag: VERB, Score: 0.999826


[('▁endors-', 'VERB'), ('toi', 'PPOBJMS'), (',', 'PUNCT'), ('▁le', 'DETMS'), ('▁ciel', 'NMS'), ('▁t', 'PREFS'), ('’', 'ADV'), ('aidera', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'MOTINC', 'score': 0.22081631, 'index': 1, 'word': '▁bouge', 'start': None, 'end': None}]


[{'entity': 'NFS', 'score': 0.72627956, 'index': 1, 'word': '▁aide', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb4

Token: ▁aide, Tag: VERB, Score: 0.999789
Token: -, Tag: VERB, Score: 0.996155
Token: toi, Tag: PPOBJMS, Score: 0.998784
Token: ,, Tag: PUNCT, Score: 0.999907
Token: ▁le, Tag: DETMS, Score: 0.999865
Token: ▁ciel, Tag: NMS, Score: 0.999726
Token: ▁t, Tag: PREFS, Score: 0.988082
Token: ’, Tag: PPOBJMP, Score: 0.202223
Token: aura, Tag: VERB, Score: 0.934380


[('▁aide-', 'VERB'), ('toi', 'PPOBJMS'), (',', 'PUNCT'), ('▁le', 'DETMS'), ('▁ciel', 'NMS'), ('▁t', 'PREFS'), ('’', 'PPOBJMP'), ('aura', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'CHIF', 'score': 0.22594386, 'index': 1, 'word': '▁aidera', 'start': None, 'end': None}]


[{'entity': 'NFS', 'score': 0.72627956, 'index': 1, 'word': '▁aide', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb5

Token: ▁ce, Tag: PRON, Score: 0.780149
Token: ▁que, Tag: PREL, Score: 0.999371
Token: ▁femme, Tag: NFS, Score: 0.999487
Token: ▁dit, Tag: VERB, Score: 0.999668
Token: ,, Tag: PUNCT, Score: 0.999905
Token: ▁dieu, Tag: PROPN, Score: 0.515889
Token: ▁le, Tag: PPOBJMS, Score: 0.998610
Token: ▁veut, Tag: VERB, Score: 0.999793


[('▁ce', 'PRON'), ('▁que', 'PREL'), ('▁femme', 'NFS'), ('▁dit', 'VERB'), (',', 'PUNCT'), ('▁dieu', 'PROPN'), ('▁le', 'PPOBJMS'), ('▁veut', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'VPPMS', 'score': 0.9966618, 'index': 1, 'word': '▁dit', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.9986129, 'index': 1, 'word': '▁veut', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb6

Token: ▁ce, Tag: PRON, Score: 0.708810
Token: ▁que, Tag: PREL, Score: 0.999433
Token: ▁femme, Tag: NFS, Score: 0.999415
Token: ▁veut, Tag: VERB, Score: 0.999820
Token: ,, Tag: PUNCT, Score: 0.999900
Token: ▁dieu, Tag: PROPN, Score: 0.934523
Token: ▁le, Tag: PPOBJMS, Score: 0.998663
Token: ▁souhaite, Tag: VERB, Score: 0.999818


[('▁ce', 'PRON'), ('▁que', 'PREL'), ('▁femme', 'NFS'), ('▁veut', 'VERB'), (',', 'PUNCT'), ('▁dieu', 'PROPN'), ('▁le', 'PPOBJMS'), ('▁souhaite', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'VPPMS', 'score': 0.9966618, 'index': 1, 'word': '▁dit', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.9986129, 'index': 1, 'word': '▁veut', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb7

Token: ▁bien, Tag: ADV, Score: 0.998597
Token: ▁mal, Tag: ADV, Score: 0.998178
Token: ▁acquis, Tag: ADJ, Score: 0.994800
Token: ▁ne, Tag: ADV, Score: 0.999753
Token: ▁sait, Tag: VERB, Score: 0.999813
Token: ▁jamais, Tag: ADV, Score: 0.999790


[('▁bien', 'ADV'), ('▁mal', 'ADV'), ('▁acquis', 'ADJ'), ('▁ne', 'ADV'), ('▁sait', 'VERB'), ('▁jamais', 'ADV')]
The form for the verb in the wordlist:

[{'entity': 'XFAMIL', 'score': 0.118672855, 'index': 1, 'word': '▁profite', 'start': None, 'end': None}]


[{'entity': 'NMS', 'score': 0.9157817, 'index': 1, 'word': '▁fait', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb8

Token: ▁bon, Tag: ADJMS, Score: 0.999459
Token: ▁ouvrier, Tag: NMS, Score: 0.999736
Token: ▁ne, Tag: ADV, Score: 0.999799
Token: ▁déplace, Tag: VERB, Score: 0.999839
Token: ▁pas, Tag: ADV, Score: 0.999800
Token: ▁ses, Tag: DET, Score: 0.999873
Token: ▁outils, Tag: NMP, Score: 0.999587


[('▁bon', 'ADJMS'), ('▁ouvrier', 'NMS'), ('▁ne', 'ADV'), ('▁déplace', 'VERB'), ('▁pas', 'ADV'), ('▁ses', 'DET'), ('▁outils', 'NMP')]
The form for the verb in the wordlist:

[{'entity': 'NMS', 'score': 0.9157817, 'index': 1, 'word': '▁fait', 'start': None, 'end': None}]


[{'entity': 'NFS', 'score': 0.99849856, 'index': 1, 'word': '▁querelle', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb9

Token: ▁pour, Tag: PREP, Score: 0.999872
Token: ▁le, Tag: DETMS, Score: 0.999878
Token: ▁fou, Tag: NMS, Score: 0.999760
Token: ,, Tag: PUNCT, Score: 0.999907
Token: ▁c, Tag: PDEMMS, Score: 0.999344
Token: ’, Tag: ADV, Score: 0.782527
Token: était, Tag: AUX, Score: 0.995531
Token: ▁tous, Tag: ADJMP, Score: 0.999119
Token: ▁les, Tag: DET, Score: 0.999871
Token: ▁jours, Tag: NMP, Score: 0.999619
Token: ▁fête, Tag: NFS, Score: 0.999738


[('▁pour', 'PREP'), ('▁le', 'DETMS'), ('▁fou', 'NMS'), (',', 'PUNCT'), ('▁c', 'PDEMMS'), ('’', 'ADV'), ('était', 'AUX'), ('▁tous', 'ADJMP'), ('▁les', 'DET'), ('▁jours', 'NMP'), ('▁fête', 'NFS')]
The form for the verb in the wordlist:

[{'entity': 'AUX', 'score': 0.9976525, 'index': 1, 'word': '▁est', 'start': None, 'end': None}]


[{'entity': 'DET', 'score': 0.3218906, 'index': 1, 'word': '▁es', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb10

Token: ▁dire, Tag: VERB, Score: 0.999755
Token: ▁et, Tag: COCO, Score: 0.999805
Token: ▁plaire, Tag: VERB, Score: 0.999709
Token: ,, Tag: PUNCT, Score: 0.999909
Token: ▁sont, Tag: AUX, Score: 0.999753
Token: ▁deux, Tag: CHIF, Score: 0.999610


[('▁dire', 'VERB'), ('▁et', 'COCO'), ('▁plaire', 'VERB'), (',', 'PUNCT'), ('▁sont', 'AUX'), ('▁deux', 'CHIF')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.99525404, 'index': 1, 'word': '▁dire', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.99781525, 'index': 1, 'word': '▁faire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb11

Token: ▁manger, Tag: VERB, Score: 0.999788
Token: ▁et, Tag: COCO, Score: 0.999811
Token: ▁faire, Tag: VERB, Score: 0.999797
Token: ,, Tag: PUNCT, Score: 0.999909
Token: ▁sont, Tag: AUX, Score: 0.999759
Token: ▁deux, Tag: CHIF, Score: 0.999605


[('▁manger', 'VERB'), ('▁et', 'COCO'), ('▁faire', 'VERB'), (',', 'PUNCT'), ('▁sont', 'AUX'), ('▁deux', 'CHIF')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.99525404, 'index': 1, 'word': '▁dire', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.99781525, 'index': 1, 'word': '▁faire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb12

Token: ▁mieux, Tag: ADV, Score: 0.989379
Token: ▁vaut, Tag: VERB, Score: 0.999807
Token: ▁prévenir, Tag: VERB, Score: 0.999780
Token: ▁que, Tag: COSUB, Score: 0.999413
Token: ▁courir, Tag: VERB, Score: 0.999823


[('▁mieux', 'ADV'), ('▁vaut', 'VERB'), ('▁prévenir', 'VERB'), ('▁que', 'COSUB'), ('▁courir', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.94116044, 'index': 1, 'word': '▁prévenir', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.9972518, 'index': 1, 'word': '▁guérir', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb13

Token: ▁mieux, Tag: ADV, Score: 0.998039
Token: ▁vaut, Tag: VERB, Score: 0.999811
Token: ▁dormir, Tag: VERB, Score: 0.999821
Token: ▁que, Tag: COSUB, Score: 0.999419
Token: ▁guérir, Tag: VERB, Score: 0.999805


[('▁mieux', 'ADV'), ('▁vaut', 'VERB'), ('▁dormir', 'VERB'), ('▁que', 'COSUB'), ('▁guérir', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.94116044, 'index': 1, 'word': '▁prévenir', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.9972518, 'index': 1, 'word': '▁guérir', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb14

Token: ▁à, Tag: PREP, Score: 0.999850
Token: ▁qui, Tag: PREL, Score: 0.988096
Token: ▁dieu, Tag: PROPN, Score: 0.964717
Token: ▁aide, Tag: VERB, Score: 0.999785
Token: ,, Tag: PUNCT, Score: 0.999903
Token: ▁nul, Tag: PRON, Score: 0.997730
Token: ▁ne, Tag: ADV, Score: 0.999759
Token: ▁peut, Tag: VERB, Score: 0.999744
Token: ▁être, Tag: AUX, Score: 0.810117


[('▁à', 'PREP'), ('▁qui', 'PREL'), ('▁dieu', 'PROPN'), ('▁aide', 'VERB'), (',', 'PUNCT'), ('▁nul', 'PRON'), ('▁ne', 'ADV'), ('▁peut', 'VERB'), ('▁être', 'AUX')]
The form for the verb in the wordlist:

[{'entity': 'NFS', 'score': 0.72627956, 'index': 1, 'word': '▁aide', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.46866497, 'index': 1, 'word': '▁nuire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb15

Token: ▁à, Tag: PREP, Score: 0.999872
Token: ▁qui, Tag: PREL, Score: 0.996137
Token: ▁dieu, Tag: PROPN, Score: 0.954418
Token: ▁veut, Tag: VERB, Score: 0.999818
Token: ,, Tag: PUNCT, Score: 0.999905
Token: ▁nul, Tag: PRON, Score: 0.997730
Token: ▁ne, Tag: ADV, Score: 0.999789
Token: ▁peut, Tag: VERB, Score: 0.999811
Token: ▁nuire, Tag: VERB, Score: 0.999732


[('▁à', 'PREP'), ('▁qui', 'PREL'), ('▁dieu', 'PROPN'), ('▁veut', 'VERB'), (',', 'PUNCT'), ('▁nul', 'PRON'), ('▁ne', 'ADV'), ('▁peut', 'VERB'), ('▁nuire', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'NFS', 'score': 0.72627956, 'index': 1, 'word': '▁aide', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.46866497, 'index': 1, 'word': '▁nuire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb16

Token: ▁il, Tag: PPER3MS, Score: 0.999731
Token: ▁faut, Tag: VERB, Score: 0.999835
Token: ▁le, Tag: PPOBJMS, Score: 0.998742
Token: ▁faire, Tag: VERB, Score: 0.999778
Token: ▁pour, Tag: PREP, Score: 0.999875
Token: ▁le, Tag: PPOBJMS, Score: 0.998742
Token: ▁croire, Tag: VERB, Score: 0.999783


[('▁il', 'PPER3MS'), ('▁faut', 'VERB'), ('▁le', 'PPOBJMS'), ('▁faire', 'VERB'), ('▁pour', 'PREP'), ('▁le', 'PPOBJMS'), ('▁croire', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.97488344, 'index': 1, 'word': '▁voir', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.9910121, 'index': 1, 'word': '▁boire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb17

Token: ▁il, Tag: PPER3MS, Score: 0.999739
Token: ▁faut, Tag: VERB, Score: 0.999838
Token: ▁le, Tag: PPOBJMS, Score: 0.998762
Token: ▁voir, Tag: VERB, Score: 0.999813
Token: ▁pour, Tag: PREP, Score: 0.999889
Token: ▁le, Tag: PPOBJMS, Score: 0.998757
Token: ▁saisir, Tag: VERB, Score: 0.999826


[('▁il', 'PPER3MS'), ('▁faut', 'VERB'), ('▁le', 'PPOBJMS'), ('▁voir', 'VERB'), ('▁pour', 'PREP'), ('▁le', 'PPOBJMS'), ('▁saisir', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.9910121, 'index': 1, 'word': '▁boire', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.99570763, 'index': 1, 'word': '▁croire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb18

Token: ▁on, Tag: PINDMS, Score: 0.998933
Token: ▁ne, Tag: ADV, Score: 0.999799
Token: ▁mord, Tag: VERB, Score: 0.999837
Token: ▁pas, Tag: ADV, Score: 0.999803
Token: ▁le, Tag: DETMS, Score: 0.999882
Token: ▁poisson, Tag: NMS, Score: 0.999760
Token: ▁qui, Tag: PREL, Score: 0.999626
Token: ▁est, Tag: VERB, Score: 0.999824
Token: ▁encore, Tag: ADV, Score: 0.999796
Token: ▁dans, Tag: PREP, Score: 0.999893
Token: ▁la, Tag: DETFS, Score: 0.999869
Token: ▁mer, Tag: NFS, Score: 0.999624


[('▁on', 'PINDMS'), ('▁ne', 'ADV'), ('▁mord', 'VERB'), ('▁pas', 'ADV'), ('▁le', 'DETMS'), ('▁poisson', 'NMS'), ('▁qui', 'PREL'), ('▁est', 'VERB'), ('▁encore', 'ADV'), ('▁dans', 'PREP'), ('▁la', 'DETFS'), ('▁mer', 'NFS')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.999252, 'index': 1, 'word': '▁vend', 'start': None, 'end': None}]


[{'entity': 'VPPMS', 'score': 0.9966618, 'index': 1, 'word': '▁dit', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb19

Token: ▁on, Tag: PINDMS, Score: 0.998933
Token: ▁ne, Tag: ADV, Score: 0.999800
Token: ▁vend, Tag: VERB, Score: 0.999836
Token: ▁pas, Tag: ADV, Score: 0.999802
Token: ▁le, Tag: DETMS, Score: 0.999878
Token: ▁poisson, Tag: NMS, Score: 0.999757
Token: ▁qui, Tag: PREL, Score: 0.999630
Token: ▁navigue, Tag: VERB, Score: 0.999827
Token: ▁encore, Tag: ADV, Score: 0.999795
Token: ▁dans, Tag: PREP, Score: 0.999894
Token: ▁la, Tag: DETFS, Score: 0.999869
Token: ▁mer, Tag: NFS, Score: 0.999648


[('▁on', 'PINDMS'), ('▁ne', 'ADV'), ('▁vend', 'VERB'), ('▁pas', 'ADV'), ('▁le', 'DETMS'), ('▁poisson', 'NMS'), ('▁qui', 'PREL'), ('▁navigue', 'VERB'), ('▁encore', 'ADV'), ('▁dans', 'PREP'), ('▁la', 'DETFS'), ('▁mer', 'NFS')]
The form for the verb in the wordlist:

[{'entity': 'AUX', 'score': 0.9976525, 'index': 1, 'word': '▁est', 'start': None, 'end': None}]


[{'entity': 'NFS', 'score': 0.99955684, 'index': 1, 'word': '▁nage', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb20

Token: ▁le, Tag: DETMS, Score: 0.999881
Token: ▁poisson, Tag: NMS, Score: 0.999758
Token: ▁mange, Tag: VERB, Score: 0.999827
Token: ▁par, Tag: PREP, Score: 0.999890
Token: ▁la, Tag: DETFS, Score: 0.999863
Token: ▁tête, Tag: NFS, Score: 0.999715


[('▁le', 'DETMS'), ('▁poisson', 'NMS'), ('▁mange', 'VERB'), ('▁par', 'PREP'), ('▁la', 'DETFS'), ('▁tête', 'NFS')]
The form for the verb in the wordlist:

[{'entity': 'ADJ', 'score': 0.5522264, 'index': 1, 'word': '▁pour', 'start': None, 'end': None}, {'entity': 'ADJMS', 'score': 0.87619734, 'index': 2, 'word': 'rit', 'start': None, 'end': None}]


[{'entity': 'CHIF', 'score': 0.22051159, 'index': 1, 'word': '▁respire', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb21

Token: ▁repose, Tag: VERB, Score: 0.999830
Token: -, Tag: VERB, Score: 0.999810
Token: toi, Tag: PPOBJMS, Score: 0.991416
Token: ▁plutôt, Tag: ADV, Score: 0.999809
Token: ▁sans, Tag: PREP, Score: 0.999899
Token: ▁souper, Tag: VERB, Score: 0.890988
Token: ,, Tag: PUNCT, Score: 0.999908
Token: ▁que, Tag: COSUB, Score: 0.999373
Token: ▁de, Tag: PREP, Score: 0.999899
Token: ▁te, Tag: PREFS, Score: 0.990941
Token: ▁lever, Tag: VERB, Score: 0.999834
Token: ▁avec, Tag: PREP, Score: 0.999898
Token: ▁des, Tag: DET, Score: 0.999872
Token: ▁de, Tag: NFP, Score: 0.999648
Token: ttes, Tag: NFP, Score: 0.999632


[('▁repose-', 'VERB'), ('toi', 'PPOBJMS'), ('▁plutôt', 'ADV'), ('▁sans', 'PREP'), ('▁souper', 'VERB'), (',', 'PUNCT'), ('▁que', 'COSUB'), ('▁de', 'PREP'), ('▁te', 'PREFS'), ('▁lever', 'VERB'), ('▁avec', 'PREP'), ('▁des', 'DET'), ('▁de', 'NFP'), ('ttes', 'NFP')]
The form for the verb in the wordlist:

[{'entity': 'NMS', 'score': 0.15039012, 'index': 1, 'word': '▁lève', 

<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb22

Token: ▁couche, Tag: VERB, Score: 0.999828
Token: -, Tag: VERB, Score: 0.999825
Token: toi, Tag: PPOBJMS, Score: 0.989640
Token: ▁plutôt, Tag: ADV, Score: 0.999810
Token: ▁sans, Tag: PREP, Score: 0.999899
Token: ▁souper, Tag: VERB, Score: 0.980420
Token: ,, Tag: PUNCT, Score: 0.999908
Token: ▁que, Tag: COSUB, Score: 0.999323
Token: ▁de, Tag: PREP, Score: 0.999898
Token: ▁te, Tag: PREFS, Score: 0.991218
Token: ▁retrouver, Tag: VERB, Score: 0.999831
Token: ▁avec, Tag: PREP, Score: 0.999898
Token: ▁des, Tag: DET, Score: 0.999872
Token: ▁de, Tag: NFP, Score: 0.999648
Token: ttes, Tag: NFP, Score: 0.999633


[('▁couche-', 'VERB'), ('toi', 'PPOBJMS'), ('▁plutôt', 'ADV'), ('▁sans', 'PREP'), ('▁souper', 'VERB'), (',', 'PUNCT'), ('▁que', 'COSUB'), ('▁de', 'PREP'), ('▁te', 'PREFS'), ('▁retrouver', 'VERB'), ('▁avec', 'PREP'), ('▁des', 'DET'), ('▁de', 'NFP'), ('ttes', 'NFP')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.9990289, 'index': 1, 'word': '

<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb23

Token: ▁les, Tag: DET, Score: 0.999874
Token: ▁montagnes, Tag: NFP, Score: 0.999625
Token: ▁ne, Tag: ADV, Score: 0.999801
Token: ▁se, Tag: PREF, Score: 0.999744
Token: ▁déplacent, Tag: VERB, Score: 0.999838
Token: ▁point, Tag: ADV, Score: 0.999783
Token: ,, Tag: PUNCT, Score: 0.999907
Token: ▁mais, Tag: COCO, Score: 0.999806
Token: ▁les, Tag: DET, Score: 0.999875
Token: ▁hommes, Tag: NMP, Score: 0.999628
Token: ▁se, Tag: PREF, Score: 0.999745
Token: ▁rencontrent, Tag: VERB, Score: 0.999790


[('▁les', 'DET'), ('▁montagnes', 'NFP'), ('▁ne', 'ADV'), ('▁se', 'PREF'), ('▁déplacent', 'VERB'), ('▁point', 'ADV'), (',', 'PUNCT'), ('▁mais', 'COCO'), ('▁les', 'DET'), ('▁hommes', 'NMP'), ('▁se', 'PREF'), ('▁rencontrent', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'CHIF', 'score': 0.22450078, 'index': 1, 'word': '▁rencontrent', 'start': None, 'end': None}]


[{'entity': 'CHIF', 'score': 0.20920987, 'index': 1, 'word': '▁paient', 'start': None, 'end': None}]


<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb24

Token: ▁les, Tag: DET, Score: 0.999873
Token: ▁montagnes, Tag: NFP, Score: 0.999626
Token: ▁ne, Tag: ADV, Score: 0.999801
Token: ▁se, Tag: PREF, Score: 0.999744
Token: ▁rencontrent, Tag: VERB, Score: 0.999825
Token: ▁point, Tag: ADV, Score: 0.999787
Token: ,, Tag: PUNCT, Score: 0.999907
Token: ▁mais, Tag: COCO, Score: 0.999806
Token: ▁les, Tag: DET, Score: 0.999874
Token: ▁hommes, Tag: NMP, Score: 0.999627
Token: ▁se, Tag: PREF, Score: 0.999745
Token: ▁déplacent, Tag: VERB, Score: 0.999833


[('▁les', 'DET'), ('▁montagnes', 'NFP'), ('▁ne', 'ADV'), ('▁se', 'PREF'), ('▁rencontrent', 'VERB'), ('▁point', 'ADV'), (',', 'PUNCT'), ('▁mais', 'COCO'), ('▁les', 'DET'), ('▁hommes', 'NMP'), ('▁se', 'PREF'), ('▁déplacent', 'VERB')]
The form for the verb in the wordlist:

[{'entity': 'CHIF', 'score': 0.22450078, 'index': 1, 'word': '▁rencontrent', 'start': None, 'end': None}]


[{'entity': 'CHIF', 'score': 0.22037283, 'index': 1, 'word': '▁battent', 'start': None, 'end': None}]

<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


Example of proverb25

Token: ▁étudier, Tag: VERB, Score: 0.999786
Token: ▁peu, Tag: ADV, Score: 0.999801
Token: ,, Tag: PUNCT, Score: 0.999907
Token: ▁chasse, Tag: VERB, Score: 0.999833
Token: ▁beaucoup, Tag: ADV, Score: 0.999797
Token: ▁de, Tag: PREP, Score: 0.999876
Token: ▁maladies, Tag: NFP, Score: 0.999590


[('▁étudier', 'VERB'), ('▁peu', 'ADV'), (',', 'PUNCT'), ('▁chasse', 'VERB'), ('▁beaucoup', 'ADV'), ('▁de', 'PREP'), ('▁maladies', 'NFP')]
The form for the verb in the wordlist:

[{'entity': 'VERB', 'score': 0.9992255, 'index': 1, 'word': '▁manger', 'start': None, 'end': None}]


[{'entity': 'VERB', 'score': 0.99127454, 'index': 1, 'word': '▁parle', 'start': None, 'end': None}]




<ipython-input-12-74977fdf26bb>:17: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  df1['Postags'][i] = tokens_after_posttraitement


## Section 3 - Modèle et code pour corriger un proverbe

Expliquez ici comment vous procédez pour choisir la meilleure version parmi les proverbes modifiés.

In [14]:
from transformers import pipeline, AutoTokenizer, AutoModelForMaskedLM, AutoModelForTokenClassification
import random


# Load Masked Language Model (MLM)
tokenizer_mlm = AutoTokenizer.from_pretrained("roberta-base")
model_mlm = AutoModelForMaskedLM.from_pretrained("roberta-base")
pipe_mlm = pipeline("fill-mask", model=model_mlm, tokenizer=tokenizer_mlm)

tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/481 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

In [15]:
from transformers import pipeline
import pandas as pd

def process_item_with_best_fit(dataframe):
    results = []

    # Get the mask token from the tokenizer to ensure it's correct
    mask_token = pipe_mlm.tokenizer.mask_token
    print(f"Mask token used: {mask_token}")

    # Convert DataFrame rows to dictionaries for easier handling
    data = dataframe.to_dict(orient="records")

    correct_count = 0  # Counter for correct predictions

    for item in data:
        # Extract the masked sentence, word list, and postags
        masked_sentence = item['Masked']
        word_list = item['Word_list']
        postags = item['Postags']  # Already in tuple format

        # Step 1: Find indices of verbs in the postags
        verb_indices = [
            i for i, (_, pos) in enumerate(postags) if "VERB" in pos
        ]

        best_final_sentence = None
        best_score = float('-inf')  # Start with a very low score

        for verb_index in verb_indices:
            # Find the verb token from postags
            verb_token = postags[verb_index][0]  # Get the token from Postags
            verb_token_clean = verb_token.lstrip("▁")  # Adjust based on tokenizer

            # Replace the verb with the mask token
            masked_sentence_with_mask = masked_sentence.replace(verb_token_clean, mask_token, 1)

            for word in word_list:
                # Replace the mask with the candidate word
                sentence_with_candidate = masked_sentence_with_mask.replace(mask_token, word)

                # Use the MLM pipeline to calculate the score of the sentence
                predictions = pipe_mlm(masked_sentence_with_mask)
                for pred in predictions:
                    if pred['token_str'].strip() == word.strip():
                        word_score = pred['score']
                        break
                else:
                    word_score = 0  # Default if word not found in predictions

                # Compare scores to select the best fit
                if word_score > best_score:
                    best_score = word_score
                    best_final_sentence = sentence_with_candidate

        # Evaluate if the best final sentence matches the original proverb
        original_proverb = item['Proverb']
        is_correct = original_proverb == best_final_sentence
        item['Final_Proverb'] = best_final_sentence
        item['Is_Correct'] = is_correct

        # Update the correct count if the prediction is correct
        if is_correct:
            correct_count += 1

        # Append the processed item to the results list
        results.append(item)

    # Calculate the percentage of correct predictions
    total_items = len(data)
    accuracy_percentage = (correct_count / total_items) * 100 if total_items > 0 else 0
    print(f"Percentage of correct predictions: {accuracy_percentage:.2f}%")

    return results, accuracy_percentage

# Assuming df1 is your DataFrame
results, accuracy = process_item_with_best_fit(df1)

# Display results and accuracy
results_df = pd.DataFrame(results)
print(results_df[['Masked', 'Proverb', 'Final_Proverb', 'Is_Correct']])
print(f"Accuracy: {accuracy:.2f}%")


Mask token used: <mask>
Percentage of correct predictions: 34.62%
                                               Masked  \
0                      a beau mentir qui part de loin   
1                     a beau dormir qui vient de loin   
2                          l’occasion forge le larron   
3                        endors-toi, le ciel t’aidera   
4                            aide-toi, le ciel t’aura   
5                      ce que femme dit, dieu le veut   
6                 ce que femme veut, dieu le souhaite   
7                      bien mal acquis ne sait jamais   
8               bon ouvrier ne déplace pas ses outils   
9            pour le fou, c’était tous les jours fête   
10                          dire et plaire, sont deux   
11                         manger et faire, sont deux   
12                     mieux vaut prévenir que courir   
13                       mieux vaut dormir que guérir   
14                  à qui dieu aide, nul ne peut être   
15                 à q

# Analyse du Code

## Chargement et utilisation du modèle

### Masked Language Model (MLM)
Le modèle `roberta-base` est utilisé comme Masked Language Model (MLM). Ce modèle permet de prédire les tokens masqués dans une phrase en se basant sur le contexte environnant. La méthode `pipeline` de Hugging Face simplifie l’interaction avec le modèle, en initialisant un pipeline spécifique pour les tâches de MLM.

## Postagging
Nous avons utilisé `qanastek/pos-french-camembert` comme modèle de postagging. Ce modèle basé sur CamemBERT fournit des balises grammaticales précises, notamment pour les verbes, identifiés principalement comme `VERB` (verbes principaux) ou `AUX` (auxiliaires).

### Observations principales :
#### Segmentation des verbes :
- CamemBERT segmente certains verbes en plusieurs tokens, chacun ayant le tag `VERB`.
- Exemple : "en", "dor", "s" → Trois tokens distincts avec le tag `VERB`.

#### Erreur de segmentation :
- Certains cas comme `-` sont identifiés à tort comme des verbes.
- Exemple : "aide – toi" → Résultat : ["aide-", "toi"].

## Traitement des données

### Post-traitement des résultats de postagging
Pour résoudre les problèmes mentionnés, un post-traitement a été mis en place :

1. **Concaténation des tokens successifs :**
   - Lorsque plusieurs tokens consécutifs ont le tag `VERB`, ils sont concaténés pour former un verbe complet.
   - Exemple : ["en", "dor", "s"] → Verbe corrigé : "endors".

2. **Gestion des cas de verbes multiples :**
   - Lorsqu’un token `VERB` est suivi d’un autre, mais que les deux représentent des verbes distincts :
     - Exemple : "vaut prévenir" ne doit pas devenir "vautprévenir".
   - Une fonction `verifie_verbe` a été mise en place pour détecter les terminaisons des trois groupes de conjugaison en français, séparant ainsi correctement ces cas.

### Résultats du post-traitement
- Ces étapes ont permis de garantir une précision de **100% dans la détection des verbes** après post-traitement.
- Un problème mineur persiste lorsque le token `-` est mal interprété comme un verbe.
  - Exemple : "aide – toi" donne "aide-" comme verbe détecté, ce qui reste acceptable car le verbe principal est bien identifié.

## Analyse des résultats

### Précision globale
  - Le modèle combiné atteint une précision de **34.62%**, indiquant des limites dans la gestion contextuelle et syntaxique des proverbes.

### Exemples d'échecs
#### Proverbes complexes :
  - **Original :** "mieux vaut prévenir que guérir"
  - **Reconstruit :** "mieux prévenir prévenir que courir".

#### Proverbes figuratifs :
  - **Original :** "a beau mentir qui vient de loin"
  - **Reconstruit :** "vient beau mentir qui part de loin".

### Exemples de succès
#### Proverbes simples :
  - **Original :** "l’occasion fait le larron"
  - **Reconstruit :** Correctement.

#### Proverbes standards :
  - **Original :** "bien mal acquis ne profite jamais"
  - **Reconstruit :** Correctement.

## Problèmes identifiés
### Limites du modèle :
  - `roberta-base` n’est pas optimisé pour des proverbes français ou des contextes figuratifs complexes.

### Traitement des candidats :
  - Certains mots pertinents sont ignorés par le modèle s’ils n’apparaissent pas dans les prédictions principales.

### Structures grammaticales complexes :
  - Les proverbes contenant plusieurs verbes ou une syntaxe non standard posent des défis.

## Propositions d'amélioration
  1. **Utilisation d’un modèle français :**
   - Remplacer `roberta-base` par CamemBERT pour mieux gérer les structures grammaticales spécifiques au français.

  2. **Enrichir le traitement des candidats :**
   - Ajouter une étape de validation grammaticale et sémantique pour améliorer la cohérence des phrases reconstruites.

  3. **Combinaison de règles linguistiques et modèles :**
   - Complémenter les prédictions avec des règles basées sur la grammaire pour éviter des incohérences syntaxiques.

  4. **Amélioration des métriques d’évaluation :**
   - Intégrer une mesure de similarité sémantique entre les proverbes reconstruits et les proverbes originaux.


## Comparaison avec la tâche 2 du TP1

### Modèle 1 : Postagging
- Dans le TP1, le postagging effectué à l’aide de SpaCy (30% de performance) est nettement inférieur à celui de CamemBERT. SpaCy n’est pas toujours correct, notamment pour des verbes conjugués comme "mord", souvent considérés comme des noms ou adjectifs.
- En revanche, le modèle Transformer (`qanastek/pos-french-camembert`) a pu bien repérer ces cas.
- Cette différence peut s’expliquer par la nature bidirectionnelle de CamemBERT (issu de BERT), qui analyse le contexte des mots à la fois à gauche et à droite pour mieux comprendre leur signification dans une phrase. Contrairement à SpaCy, qui ne prend en compte que le contexte précédent (c’est-à-dire à gauche), ce qui limite sa capacité à effectuer un postagging précis.

### Modèle 2 : Correction des verbes
- Dans le TP1, la correction basée sur un modèle de masquage des verbes était moins performante, avec un score de masquage de **21%**.
- Notre modèle Transformer (`roberta-base`) atteint une précision de **34%**, soit une amélioration de **13%**.
- Cette différence est attribuée à la capacité de `roberta-base` à utiliser le contexte à gauche et à droite pour mieux prédire les verbes masqués.
- De plus, la meilleure performance du modèle de postagging (Modèle 1) améliore la détection des verbes dans les phrases, renforçant ainsi la correction des verbes.